validator_test.go

```diff
@@ -0,0 +1,361 @@
+	values := make(url.Values)
+	for k, v := range m {
+		values.Set(k, v)
+	}
+	return values
+}
+
+func makeRequest(method, url, authHeader string, values url.Values) *http.Request {
+	req := httptest.NewRequest(method, url, strings.NewReader(values.Encode()))
+	if authHeader != "" {
+		req.Header.Set(authorizationHeaderParam, authHeader)
+	}
+	req.Header.Set(contentType, formContentType)
+	return req
+}
+
+func TestNewProviderRequest_ParamSources(t *testing.T) {
+	expectedParams := map[string]string{
+		"oauth_consumer_key":     "some_client",
+		"oauth_signature":        "sig",
+		"oauth_nonce":            "a_nonce",
+		"oauth_timestamp":        "1",
+		"oauth_signature_method": "HMAC-SHA1",
+	}
+	expectedValues := makeValues(expectedParams)
+	url := "https://example.com/oauth1"
+	queryReq := makeRequest("GET", url+"?"+expectedValues.Encode(), "", nil)
+	bodyReq := makeRequest("POST", url, "", expectedValues)
+	headerReq := makeRequest("GET", url, authHeaderValue(expectedParams), nil)
+
+	delete(expectedParams, "oauth_signature")
+
+	for i, req := range []*http.Request{queryReq, bodyReq, headerReq} {
+		preq, err := newProviderRequest(req)
+		assert.NoError(t, err, i)
+		assert.Equal(t, expectedParams, preq.oauthParams, i)
+		assert.Same(t, req, preq.req, i)
+		assert.Equal(t, preq.signatureToVerify, "sig", i)
+		assert.Equal(t, preq.signatureMethod, "HMAC-SHA1", i)
+		assert.Equal(t, preq.nonce, "a_nonce", i)
+		assert.Equal(t, preq.timestamp, int64(1), i)
+		assert.Equal(t, preq.clientKey, "some_client", i)
+
+	}
+}
+
+func TestNewProviderRequest_CombinedParams(t *testing.T) {
+	req := makeRequest("POST", "https://example.com/oauth?query=1", "OAUTH header=\"1\"", makeValues(map[string]string{
+		"body":                   "1",
+		"oauth_consumer_key":     "some_client",
+		"oauth_signature":        "sig",
+		"oauth_nonce":            "a_nonce",
+		"oauth_timestamp":        "1",
+		"oauth_signature_method": "HMAC-SHA1",
+	}))
+	preq, err := newProviderRequest(req)
+	assert.Nil(t, err)
+	assert.Equal(t, map[string]string{
+		"header":                 "1",
+		"query":                  "1",
+		"body":                   "1",
+		"oauth_consumer_key":     "some_client",
+		"oauth_nonce":            "a_nonce",
+		"oauth_timestamp":        "1",
+		"oauth_signature_method": "HMAC-SHA1",
+	}, preq.oauthParams)
+}
+
+func TestNewProviderRequest_InvalidAuthHeader(t *testing.T) {
+	req := makeRequest("POST", "https://example.com/oauth", "OAUTH thisisn'tvalid", makeValues(map[string]string{
+		"oauth_consumer_key":     "some_client",
+		"oauth_signature":        "sig",
+		"oauth_nonce":            "a_nonce",
+		"oauth_timestamp":        "1",
+		"oauth_signature_method": "HMAC-SHA1",
+	}))
+	_, err := newProviderRequest(req)
+	assert.NotNil(t, err)
+}
+
+func TestNewProviderRequest_OtherAuthHeader(t *testing.T) {
+	req := makeRequest("POST", "https://example.com/oauth", "bearer foobar", makeValues(map[string]string{
```

Comment:
Why is this allowed? This feels like an invalid request, but the test is clearly letting it be created. Is it just that `ValidateSignature()` would fail, but `newProviderRequest()` wouldn't?

Reply:
Right. Basically I'm testing that newProviderRequest ignores the Authorization header if it isn't using oauth.
ValidateSignature would fail since there's no signature in the query or form body.